packages/protocol/contracts/dapis/SamplePriceDataFeed.sol

```diff
@@ -19,6 +37,170 @@ contract SamplePriceDataFeed {
     latestDapiId = dapiId;
   }
 
+  function addTemplate(
+    bytes16 _dapiId,
+    bytes32 _templateId,
+    address _designatedWallet
+  ) external onlyAdminOrMetaAdmin cooldownTimeEnded {
+    // PriceDataFeed fetches the current dAPI parameters from RrpDapiServer
+    (
+      uint256 noResponsesToReduce,
+      uint256 toleranceInPercentages,
+      uint256 requesterIndex,
+      bytes32[] memory templateIds,
+      address[] memory designatedWallets,
+      address reduceAddress,
+      bytes4 reduceFunctionId,
+      address requestIndexResetter
+    ) = rrpDapiServer.getDapi(_dapiId);
+    // TODO: check dAPI was found?
+
+    // Adds templateId to dAPI.templateIds (iterating over because array push() is only available in storage arrays)
+    // Also add _designatedWallet to dAPI.designatedWallets
+    bytes32[] memory newTemplateIds = new bytes32[](templateIds.length + 1);
+    address[] memory newDesignatedWallets = new address[](designatedWallets.length + 1);
+    for (uint256 i = 0; i < templateIds.length; i++) {
+      newTemplateIds[i] = templateIds[i];
+      newDesignatedWallets[i] = designatedWallets[i];
+    }
+    newTemplateIds[newTemplateIds.length - 1] = _templateId;
+    newDesignatedWallets[newDesignatedWallets.length - 1] = _designatedWallet;
+
+    // Calls RrpDapiServer to create the new dAPI
+    bytes16 newDapiId =
+      rrpDapiServer.registerDapi(
+        noResponsesToReduce,
+        toleranceInPercentages,
+        requesterIndex,
+        newTemplateIds,
+        newDesignatedWallets,
+        reduceAddress,
+        reduceFunctionId,
+        requestIndexResetter
+      );
+
+    // Update dapiLastUpdated timestamp
+    adminToDapiLastUpdated[msg.sender] = uint64(block.timestamp);
+
+    // Record the returned dapiId
+    latestDapiId = newDapiId;
+
+    // TODO: publish TemplateAdded event?
+  }
+
+  function removeTemplate(bytes16 _dapiId, bytes32 _templateId) external onlyAdminOrMetaAdmin cooldownTimeEnded {
+    // PriceDataFeed fetches the current dAPI parameters from RrpDapiServer
+    (
+      uint256 noResponsesToReduce,
+      uint256 toleranceInPercentages,
+      uint256 requesterIndex,
+      bytes32[] memory templateIds,
+      address[] memory designatedWallets,
+      address reduceAddress,
+      bytes4 reduceFunctionId,
+      address requestIndexResetter
+    ) = rrpDapiServer.getDapi(_dapiId);
```

Comment:
I'm actually not familiar with this notation ...

Reply:
I see this notation to something similar to [destructuring](https://docs.soliditylang.org/en/v0.8.4/control-structures.html#destructuring-assignments-and-returning-multiple-values) in Typescript. 

I would have preferred to be able to just return a struct or an some other data structure from RrpDapiServer.getDapi() but I found out that it was not possible therefore that's why I ended up returning all the individual values.

Reply:
ah, yes yes. same thing in python ("unpacking" values). Yea, this looks nice.